pyrdp/layer/layer.py

```diff
@@ -100,41 +170,32 @@ def chain(first: 'Layer', second: 'Layer', *layers: ['Layer']):
         for nextLayer in layers:
             current.setNext(nextLayer)
             current = nextLayer
-    
-    def setNext(self, layer: 'Layer'):
+
+    def setNext(self, layer: 'BaseLayer'):
         """
         Set the next layer in the protocol hierarchy (ex: IP's next layer would be TCP/UDP).
         :param layer: The next layer.
         """
         self.next = layer
-        layer.previous = self
-    
-    def pduReceived(self, pdu: PDU, forward: bool):
-        """
-        Called when a PDU is received.
-        Notifies the attached observer and optionally forwards the payload to the next layer.
-        :param pdu: the PDU.
-        :param forward: whether the PDU's payload should be forwarded to the next layer.
-        :type forward: bool
-        """
-        self.eventEngine.processObject(pdu)
+        layer.setPrevious(self)
 
-        if self.observer is not None:
-            self.observer.onPDUReceived(pdu)
-        
-        if forward and self.next is not None:
-            self.next.recv(pdu.payload)
 
-    async def waitPDU(self, *args, **kwargs):
-        """
-        Wait for a PDU matching certain criteria (see EventEngine.wait)
-        :return: PDU
-        """
-        return await self.eventEngine.wait(*args, **kwargs)
+class IntermediateLayer(Layer, LayerChainItem):
+    """
+    Layer that usually forwards some or all PDUs to another layer.
```

Comment:
Yes, basically anything that isn't an "endpoint" layer. Slow-path and clipboard are examples of layers that are not intermediate (since all the data is consumed).

Reply:
(I'll add an example)